File: lib/csharp/microsoft.bot.builder.skills/Microsoft.Bot.Builder.Skills/SkillHttpTransport.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder.Skills.Auth;
using Microsoft.Bot.Builder.Skills.Models;
using Microsoft.Bot.Builder.Skills.Models.Manifest;
using Microsoft.Bot.Builder.Solutions;
using Microsoft.Bot.Connector.Authentication;
using Microsoft.Bot.Schema;
using Microsoft.Rest.Serialization;

namespace Microsoft.Bot.Builder.Skills
{
    public class SkillHttpTransport : ISkillTransport
    {
        private HttpClient _httpClient = new HttpClient();
        private readonly SkillManifest _skillManifest;
        private readonly MicrosoftAppCredentialsEx _microsoftAppCredentialsEx;

        // Protected to enable mocking
        protected HttpClient HttpClient { get => _httpClient; set => _httpClient = value; }

        public SkillHttpTransport(SkillManifest skillManifest, MicrosoftAppCredentialsEx microsoftAppCredentialsEx)
        {
            _skillManifest = skillManifest ?? throw new ArgumentNullException(nameof(skillManifest));
            _microsoftAppCredentialsEx = microsoftAppCredentialsEx ?? throw new ArgumentNullException(nameof(microsoftAppCredentialsEx));
        }

        public async Task CancelRemoteDialogsAsync(ITurnContext turnContext)
        {
            var cancelRemoteDialogEvent = turnContext.Activity.CreateReply();

            cancelRemoteDialogEvent.Type = ActivityTypes.Event;
            cancelRemoteDialogEvent.Name = SkillEvents.CancelAllSkillDialogsEventName;

            await ForwardToSkillAsync(turnContext, cancelRemoteDialogEvent);
        }

        public void Disconnect()
        {
            // doesn't have to do any disconnect for http
        }

        public async Task<bool> ForwardToSkillAsync(ITurnContext dialogContext, Activity activity, Func<Activity, Activity> tokenRequestHandler = null)
        {
            // Serialize the activity and POST to the Skill endpoint
            var httpRequest = new HttpRequestMessage();
            httpRequest.Method = new HttpMethod("POST");
            httpRequest.RequestUri = _skillManifest.Endpoint;

            var requestContent = SafeJsonConvert.SerializeObject(activity, Serialization.Settings);
            httpRequest.Content = new StringContent(requestContent, Encoding.UTF8);
            httpRequest.Content.Headers.ContentType = MediaTypeHeaderValue.Parse("application/json; charset=utf-8");

            MicrosoftAppCredentials.TrustServiceUrl(_skillManifest.Endpoint.AbsoluteUri);
            await _microsoftAppCredentialsEx.ProcessHttpRequestAsync(httpRequest, default(CancellationToken));

            var response = await _httpClient.SendAsync(httpRequest);

            if (response.IsSuccessStatusCode)
            {
                var skillResponses = new List<Activity>();
                var filteredSkillResponses = new List<Activity>();

                // Retrieve Activity responses
                var responseStr = await response.Content.ReadAsStringAsync();
                skillResponses = SafeJsonConvert.DeserializeObject<List<Activity>>(responseStr, Serialization.Settings);

                var endOfConversation = false;
                foreach (var skillResponse in skillResponses)
                {
                    // Once a Skill has finished it signals that it's handing back control to the parent through a
                    // EndOfConversation event which then causes the SkillDialog to be closed. Otherwise it remains "in control".
                    if (skillResponse.Type == ActivityTypes.EndOfConversation)
                    {
                        endOfConversation = true;
                    }
                    else if (skillResponse?.Name == TokenEvents.TokenRequestEventName)
                    {
                        if (tokenRequestHandler != null)
                        {
                            var tokenResponseActivity = tokenRequestHandler(skillResponse);
                            if (tokenResponseActivity != null)
                            {
                                return await ForwardToSkillAsync(dialogContext, tokenResponseActivity);
                            }
                        }
                    }
                    else
                    {
                        filteredSkillResponses.Add(skillResponse);
                    }
                }

                // Send the filtered activities back (for example, token requests, EndOfConversation, etc. are removed)
                if (filteredSkillResponses.Count > 0)
                {
                    await dialogContext.SendActivitiesAsync(filteredSkillResponses.ToArray());
                }

                return endOfConversation;
            }
            else
            {
                await dialogContext.SendActivityAsync(new Activity(type: ActivityTypes.Trace, text: $"HTTP error when forwarding activity to the skill: Status Code:{response.StatusCode}, Message:{response.ReasonPhrase}"));
                throw new HttpRequestException(response.ReasonPhrase);
            }
        }
    }
}

File: lib/typescript/botbuilder-skills/src/skillHttpTransport.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import { DefaultHttpClient, HttpClient, HttpOperationResponse, RequestPrepareOptions, WebResource } from '@azure/ms-rest-js';
import { TurnContext } from 'botbuilder';
import { ActivityExtensions, TokenEvents } from 'botbuilder-solutions';
import { Activity, ActivityTypes } from 'botframework-schema';
import { MicrosoftAppCredentialsEx } from './auth';
import { ISkillManifest, SkillEvents } from './models';
import { ISkillTransport, TokenRequestHandler } from './skillTransport';

export class SkillHttpTransport implements ISkillTransport {
    private readonly httpClient: HttpClient;
    private readonly skillManifest: ISkillManifest;
    private readonly appCredentials: MicrosoftAppCredentialsEx;

    /**
     * Http SkillTransport implementation
     */
    constructor(skillManifest: ISkillManifest, appCredentials: MicrosoftAppCredentialsEx, httpClient?: HttpClient) {
        if (!skillManifest) {
            throw new Error('skillManifest has no value');
        }
        this.skillManifest = skillManifest;

        if (!appCredentials) {
            throw new Error('appCredentials has no value');
        }
        this.appCredentials = appCredentials;

        this.httpClient = httpClient || new DefaultHttpClient();
    }

    public async forwardToSkill(turnContext: TurnContext,
                                activity: Partial<Activity>,
                                tokenRequestHandler?: TokenRequestHandler): Promise<boolean> {
        // Serialize the activity and POST to the Skill endpoint
        const requestOptions: RequestPrepareOptions = {
            method: 'POST',
            url: this.skillManifest.endpoint,
            body: activity
        };
        const request: WebResource = new WebResource().prepare(requestOptions);

        MicrosoftAppCredentialsEx.trustServiceUrl(this.skillManifest.endpoint);
        await this.appCredentials.signRequest(request);

        const response: HttpOperationResponse = await this.httpClient.sendRequest(request);

        if (response.status < 200 || response.status >= 300) {
            const result: string = `HTTP error when forwarding activity to the skill: Status Code:${
                response.status
            }, Message: '${
                response.bodyAsText
            }'`;

            await turnContext.sendActivity({
                type: ActivityTypes.Trace,
                text: result
            });

            throw new Error(result);
        }

        const responseBody: Activity[] = response.parsedBody;

        // Retrieve Activity responses
        const skillResponses: Activity[] = responseBody.map(fixActivityTimestamp);

        const filteredResponses: Activity[] = [];

        let endOfConversation: boolean = false;

        skillResponses.forEach(async (skillResponse: Activity) => {
            // Once a Skill has finished it signals that it's handing back control to the parent through a
            // EndOfConversation event which then causes the SkillDialog to be closed. Otherwise it remains "in control".
            if (skillResponse.type === ActivityTypes.EndOfConversation) {
                endOfConversation = true;
            } else if (skillResponse.name === TokenEvents.tokenRequestEventName) {
                if (tokenRequestHandler) {
                    const tokenResponseActivity: Activity|undefined = await tokenRequestHandler(skillResponse);
                    if (tokenResponseActivity) {
                        return this.forwardToSkill(turnContext, tokenResponseActivity);
                    }
                }
            } else {
                // Trace messages are not filtered out and are sent along with messages/events.
                filteredResponses.push(skillResponse);
            }
        });

        // Send the filtered activities back (for example, token requests, EndOfConversation, etc. are removed)
        if (filteredResponses.length > 0) {
            await turnContext.sendActivities(filteredResponses);
        }

        return endOfConversation;
    }

    public async cancelRemoteDialogs(turnContext: TurnContext): Promise<void> {
        const cancelRemoteDialogEvent: Activity = ActivityExtensions.createReply(turnContext.activity);
        cancelRemoteDialogEvent.type = ActivityTypes.Event;
        cancelRemoteDialogEvent.name = SkillEvents.cancelAllSkillDialogsEventName;

        await this.forwardToSkill(turnContext, cancelRemoteDialogEvent);
    }

    public disconnect(): void {
        // doesn't have to do any disconnect for http
    }
}

/**
 * Parse the Date-related properties
 * @param activity Activity to convert
 * @returns Activity with correct Date properties
 */
function fixActivityTimestamp(activity: Activity): Activity {
    if (typeof activity.timestamp === 'string') {
        activity.timestamp = new Date(activity.timestamp);
    }

    if (typeof activity.localTimestamp === 'string') {
        activity.localTimestamp = new Date(activity.localTimestamp);
    }

    if (typeof activity.expiration === 'string') {
        activity.expiration = new Date(activity.expiration);
    }

    return activity;
}
